src/windows/find_tools.rs

```diff
@@ -482,28 +493,42 @@ mod impl_ {
     ) -> Option<(PathBuf, PathBuf, PathBuf, PathBuf, Option<PathBuf>, PathBuf)> {
         let version = vs15plus_vc_read_version(instance_path)?;
 
-        let host = match host_arch() {
-            X86 => "X86",
-            X86_64 => "X64",
-            // There is no natively hosted compiler on ARM64.
-            // Instead, use the x86 toolchain under emulation (there is no x64 emulation).
-            AARCH64 => "X86",
+        let hosts = match host_arch() {
+            X86 => &["X86"],
+            X86_64 => &["X64"],
+            // Starting with VS 17.4, there is a natively hosted compiler on ARM64:
+            // https://devblogs.microsoft.com/visualstudio/arm64-visual-studio-is-officially-here/
+            // On older versions of VS, we use x64 if running under emulation is supported,
+            // otherwise use x86.
+            AARCH64 => {
+                if is_amd64_emulation_supported() {
+                    &["ARM64", "X64", "X86"][..]
+                } else {
+                    &["ARM64", "X86"]
+                }
+            }
             _ => return None,
         };
         let target = lib_subdir(target)?;
         // The directory layout here is MSVC/bin/Host$host/$target/
         let path = instance_path.join(r"VC\Tools\MSVC").join(version);
+        // We use the first available host architecture that can build for the target
+        let (host_path, host) = hosts.iter().find_map(|&x| {
+            let candidate = path.join("bin").join(&format!("Host{}", x));
+            if candidate.join(&target).exists() {
+                Some((candidate, x))
+            } else {
+                None
+            }
+        })?;
         // This is the path to the toolchain for a particular target, running
         // on a given host
-        let bin_path = path.join("bin").join(format!("Host{}", host)).join(target);
+        let bin_path = host_path.join(&target);
         // But! we also need PATH to contain the target directory for the host
         // architecture, because it contains dlls like mspdb140.dll compiled for
         // the host architecture.
-        let host_dylib_path = path
-            .join("bin")
-            .join(format!("Host{}", host))
-            .join(host.to_lowercase());
-        let lib_path = path.join("lib").join(target);
+        let host_dylib_path = host_path.join(&host.to_lowercase());
```

Comment:
No, because `host_path` is to the host directory which contains the subdirectories for each target, whereas `host_dyn_path` points to the target directory in there for the current host.

For example:
`host_path` = `...\bin\Hostx64`
`host_dyn_path` = `...\bin\Hostx64\x64`